File: q4_Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Float}
def find_median_sorted_arrays(nums1, nums2)
  len = nums1.length + nums2.length
  nums = []
  i = j = 0
  (0..len).each do |_index|
    if i < nums1.length && j < nums2.length
      if nums1[i] <= nums2[j]
        nums << nums1[i]
        i += 1
      else
        nums << nums2[j]
        j += 1
      end
    elsif i < nums1.length
      nums << nums1[i]
      i += 1
    else
      nums << nums2[j]
      j += 1
    end
  end
  (len % 2).zero? ? (nums[len / 2] + nums[len / 2 - 1]) / 2.0 : nums[len / 2] / 1.0
end

File: 0004_Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.py
# -*- coding: utf-8 -*-

'''
There are two sorted arrays nums1 and nums2 of size m and n respectively.

Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

You may assume nums1 and nums2 cannot be both empty.

Example 1:

nums1 = [1, 3]
nums2 = [2]

The median is 2.0

Example 2:

nums1 = [1, 2]
nums2 = [3, 4]

The median is (2 + 3)/2 = 2.5
'''


class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        nums = []
        length = len(nums1) + len(nums2)
        i = j = 0
        for _ in range(length):
            if i < len(nums1) and j < len(nums2):
                if nums1[i] <= nums2[j]:
                    nums.append(nums1[i])
                    i += 1
                else:
                    nums.append(nums2[j])
                    j += 1
            elif i < len(nums1):
                nums.append(nums1[i])
                i += 1
            else:
                nums.append(nums2[j])
                j += 1
        half = length // 2
        return (nums[half] + nums[~half]) / 2
